libs/providers/ofrep-web/src/lib/ofrep-web-provider.ts

```diff
@@ -0,0 +1,352 @@
+      }
+
+      const bulkSuccessResp = response.value;
+      const newCache: { [key: string]: ResolutionDetails<FlagValue> | ResolutionError } = {};
+
+      bulkSuccessResp.flags?.forEach((evalResp: EvaluationResponse) => {
+        if (isEvaluationFailureResponse(evalResp)) {
+          newCache[evalResp.key] = {
+            errorCode: evalResp.errorCode,
+            errorDetails: evalResp.errorDetails,
+            reason: StandardResolutionReasons.ERROR,
+          };
+        }
+
+        if (isEvaluationSuccessResponse(evalResp) && evalResp.key) {
+          newCache[evalResp.key] = {
+            value: evalResp.value,
+            flagMetadata: evalResp.metadata as FlagMetadata,
+            reason: evalResp.reason,
+            variant: evalResp.variant,
+          };
+        }
+      });
+      const listUpdatedFlags = this._getListUpdatedFlags(this._cache, newCache);
+      this._cache = newCache;
+      this._etag = response.httpResponse?.headers.get('etag');
+      return { status: BulkEvaluationStatus.SUCCESS_WITH_CHANGES, flags: listUpdatedFlags };
+    } catch (error) {
+      if (error instanceof OFREPApiTooManyRequestsError) {
+        this._setRetryAfter(error);
+      }
+      throw error;
+    }
+  }
+
+  /**
+   * _getListUpdatedFlags is a function that will compare the old cache with the new cache and
+   * return the list of flags that have been updated / deleted / created.
+   * @param oldCache
+   * @param newCache
+   * @private
+   */
+  private _getListUpdatedFlags(
+    oldCache: { [key: string]: ResolutionDetails<FlagValue> | ResolutionError },
+    newCache: { [key: string]: ResolutionDetails<FlagValue> | ResolutionError },
+  ): string[] {
+    const changedKeys: string[] = [];
+    const oldKeys = Object.keys(oldCache);
+    const newKeys = Object.keys(newCache);
+
+    // Check for added or modified keys in newCache
+    for (const key in newCache) {
+      if (oldKeys.indexOf(key) === -1 || JSON.stringify(oldCache[key]) !== JSON.stringify(newCache[key])) {
+        changedKeys.push(key);
+      }
+    }
+
+    // Check for removed keys in oldCache
+    for (const key in oldCache) {
+      if (newKeys.indexOf(key) === -1) {
+        changedKeys.push(key);
+      }
+    }
+
+    return changedKeys;
+  }
+
+  /**
+   * _setRetryAfter is setting the date until we should stop polling based on the error.
+   * @param error - OFREPApiTooManyRequestsError received
+   * @private
+   */
+  private _setRetryAfter(error: OFREPApiTooManyRequestsError) {
+    const retryAfter = error.response?.headers.get('Retry-After');
+    if (retryAfter) {
+      if (!isNaN(retryAfter as unknown as number)) {
+        this._retryPollingAfter = new Date(Date.now() + (retryAfter as unknown as number) * 1000);
+      } else {
+        const retryAfterDate = new Date(retryAfter);
+        if (retryAfterDate.toString() !== 'Invalid Date') {
+          this._retryPollingAfter = retryAfterDate;
+        }
+      }
+    }
+  }
```

Comment:
I guess we should not implicitly convert the string header into numbers here.
I just opened a PR to solve paring the header in ofrep-shared.
With this you should be able to just do `error.retryAfterDate ` and assign it to `this._retryPollingAfter` if it is not null.

```js
private _setRetryAfter(error: OFREPApiTooManyRequestsError) {
  const retryAfterDate = error.retryAfterDate
  if(retryAfterDate){
    this._retryPollingAfter = retryAfterDate
  }
}
```

Reply:
That's a good idea to put the logic in `ofrep-core`.
I will update the code when we will release the new version.

Reply:
It is deployed :)